File: backend/models/code/github.py
"""
GitHub integration models (Code domain).

Each Code session (``CodeProject``) maps 1:1 to a GitHub repository created under
the org where the platform's GitHub App is installed. ``GitHubRepoLink`` records
that mapping (so repeated pushes target the same repo and accrue version history);
``GitHubPushLog`` is the append-only history of sync attempts.

Neither model stores any secret — the App private key lives in the environment,
and installation tokens are short-lived and never persisted.
"""
import uuid
from datetime import datetime

from backend.extensions import db


class GitHubPushStatus:
    """Sync attempt status constants."""

    PENDING = "pending"
    SUCCESS = "success"
    FAILED = "failed"


class GitHubRepoLink(db.Model):
    """1:1 mapping from a Code session to its GitHub repository."""

    __tablename__ = "github_repo_links"

    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = db.Column(
        db.String(36), db.ForeignKey("code_projects.id"), nullable=False, index=True
    )
    user_id = db.Column(db.String(36), db.ForeignKey("users.id"), nullable=False, index=True)
    team_id = db.Column(db.String(36), db.ForeignKey("teams.id"), nullable=True, index=True)

    installation_id = db.Column(db.String(40), nullable=True)
    repo_owner = db.Column(db.String(120), nullable=False)
    repo_name = db.Column(db.String(120), nullable=False)
    repo_id = db.Column(db.String(40), nullable=True)
    default_branch = db.Column(db.String(80), nullable=False, default="main")
    html_url = db.Column(db.String(400), nullable=True)
    visibility = db.Column(db.String(20), nullable=False, default="private")

    # Denormalised latest-push summary for cheap status reads.
    last_commit_sha = db.Column(db.String(40), nullable=True)
    last_pushed_at = db.Column(db.DateTime, nullable=True)
    last_status = db.Column(db.String(20), nullable=True)

    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # One repo per session.
    __table_args__ = (db.UniqueConstraint("project_id", name="uq_github_repo_link_project"),)

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "project_id": self.project_id,
            "repo_owner": self.repo_owner,
            "repo_name": self.repo_name,
            "full_name": f"{self.repo_owner}/{self.repo_name}",
            "default_branch": self.default_branch,
            "html_url": self.html_url,
            "visibility": self.visibility,
            "last_commit_sha": self.last_commit_sha,
            "last_pushed_at": self.last_pushed_at.isoformat() + "Z"
            if self.last_pushed_at
            else None,
            "last_status": self.last_status,
            "created_at": self.created_at.isoformat() + "Z" if self.created_at else None,
            "updated_at": self.updated_at.isoformat() + "Z" if self.updated_at else None,
        }


class GitHubPushLog(db.Model):
    """A single auto-sync attempt (one per completed generation run)."""

    __tablename__ = "github_push_logs"

    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = db.Column(
        db.String(36), db.ForeignKey("code_projects.id"), nullable=False, index=True
    )
    repo_link_id = db.Column(
        db.String(36), db.ForeignKey("github_repo_links.id"), nullable=True, index=True
    )
    run_id = db.Column(db.String(36), nullable=True, index=True)

    status = db.Column(db.String(20), nullable=False, default=GitHubPushStatus.PENDING)
    branch = db.Column(db.String(80), nullable=True)
    commit_sha = db.Column(db.String(40), nullable=True)
    files_count = db.Column(db.Integer, nullable=True)
    message = db.Column(db.Text, nullable=True)
    error_message = db.Column(db.Text, nullable=True)

    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    finished_at = db.Column(db.DateTime, nullable=True)

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "project_id": self.project_id,
            "repo_link_id": self.repo_link_id,
            "run_id": self.run_id,
            "status": self.status,
            "branch": self.branch,
            "commit_sha": self.commit_sha,
            "files_count": self.files_count,
            "message": self.message,
            "error_message": self.error_message,
            "created_at": self.created_at.isoformat() + "Z" if self.created_at else None,
            "finished_at": self.finished_at.isoformat() + "Z" if self.finished_at else None,
        }

File: frontend/src/api/github.ts
/**
 * GitHub integration API client (Code domain).
 *
 * Read-only: syncing is automatic (the agent runtime pushes each session's
 * deliverables to its repo when a `code_*` run completes). These endpoints just
 * surface whether the org-level GitHub App is configured and the per-session
 * repo link + push history for status display.
 */
import { api } from "@/api/client"

export interface GitHubStatus {
  configured: boolean
  connected: boolean
  owner?: string
  installation_id?: string
  error?: string
}

export interface GitHubRepo {
  id: string
  project_id: string
  repo_owner: string
  repo_name: string
  full_name: string
  default_branch: string
  html_url: string | null
  visibility: string
  last_commit_sha: string | null
  last_pushed_at: string | null
  last_status: string | null
  created_at?: string | null
  updated_at?: string | null
  // The branch the platform forks (once) for secondary development + a clone URL.
  dev_branch?: string | null
  clone_url?: string | null
}

export interface GitHubSyncResult {
  status: string
  repo_url?: string | null
  full_name?: string
  branch?: string
  commit_sha?: string
  files?: number
  dev_branch?: string
  skipped?: { path: string; size: number }[]
  error?: string
}

export interface GitHubPush {
  id: string
  project_id: string
  run_id: string | null
  status: "pending" | "success" | "failed"
  branch: string | null
  commit_sha: string | null
  files_count: number | null
  message: string | null
  error_message: string | null
  created_at?: string | null
  finished_at?: string | null
}

export interface GitHubProjectRepo {
  configured: boolean
  linked: boolean
  repo: GitHubRepo | null
  last_push: GitHubPush | null
}

interface Envelope<T> {
  data: T
  message?: string
}

export const githubApi = {
  getStatus: async (): Promise<GitHubStatus> => {
    const response = await api.get<Envelope<GitHubStatus>>("/code/github/status")
    return response.data
  },
  getProjectRepo: async (projectId: string): Promise<GitHubProjectRepo> => {
    const response = await api.get<Envelope<GitHubProjectRepo>>(
      `/code/github/projects/${projectId}/repo`
    )
    return response.data
  },
  getProjectPushes: async (projectId: string): Promise<GitHubPush[]> => {
    const response = await api.get<Envelope<{ pushes: GitHubPush[] }>>(
      `/code/github/projects/${projectId}/pushes`
    )
    return response.data.pushes
  },
  // Idempotent self-service re-push (reuses the auto-sync path server-side).
  syncProject: async (projectId: string): Promise<GitHubSyncResult> => {
    const response = await api.post<Envelope<GitHubSyncResult>>(
      `/code/github/projects/${projectId}/sync`
    )
    return response.data
  },
}
